odoo/fields.py

```diff
@@ -273,47 +277,77 @@ def __repr__(self):
     #
     # Base field setup: things that do not depend on other models/fields
     #
-
-    def setup_base(self, model, name):
-        """ Base setup: things that do not depend on other models/fields. """
-        if self._setup_done and not self.related:
-            # optimization for regular fields: keep the base setup
-            self._setup_done = 'base'
-        else:
-            # do the base setup from scratch
-            self._setup_attrs(model, name)
-            if not self.related:
-                self._setup_regular_base(model)
-            self._setup_done = 'base'
+    # The base field setup is done by field.__set_name__(), which determines the
+    # field's name, model name, module and its parameters.
+    #
+    # The dictionary field.args gives the parameters passed to the field's
+    # constructor.  Most parameters have an attribute of the same name on the
+    # field.  The parameters as attributes are assigned by the field setup.
+    #
+    # In order to save memory, a field should avoid having field.args and/or
+    # many attributes when possible.  We call "direct" a field that can be set
+    # up directly from its definition class.  Direct fields are non-related
+    # fields defined on models, and can be shared across registries.  We call
+    # "toplevel" a field that is put on the model's registry class, and is
+    # therefore specific to the registry.
+    #
+    # Toplevel field are set up once, and are no longer set up from scratch
+    # after that.  Those fields can save memory by discarding field.args and
+    # field._base_fields once set up, because those are no longer necessary.
+    #
+    # Non-toplevel non-direct fields are the fields on definition classes that
+    # may not be shared.  In other words, those fields are never used directly,
+    # and are always recreated as toplevel fields.  On those fields, the base
+    # setup is useless, because only field.args is used for setting up other
+    # fields.  We therefore skip the base setup for those fields.  The only
+    # attributes of those fields are: '_sequence', 'args', 'model_name', 'name'
+    # and '_module', which makes their __dict__'s size minimal.
+
+    def __set_name__(self, owner, name):
```

Comment:
It's doing more than setting name

Reply:
As the docstring says, it performs the base setup of the field.

Reply:
BTW, this method is a special method of Python, see https://docs.python.org/3/reference/datamodel.html?highlight=__set_name__#object.__set_name__